protocol/messenger_communities.go

```diff
@@ -3635,3 +3654,64 @@ func (m *Messenger) CheckPermissionsToJoinCommunity(request *requests.CheckPermi
 
 	return m.communitiesManager.CheckPermissionToJoin(request.CommunityID, addresses)
 }
+
+// startCommunityRekeyLoop creates a 5-minute ticker and starts a routine that attempts to rekey every community every tick
+func (m *Messenger) startCommunityRekeyLoop() {
+	logger := m.logger.Named("CommunityRekeyLoop")
+
+	var d time.Duration
+	if m.communitiesManager.RekeyInterval != 0 {
+		d = m.communitiesManager.RekeyInterval / 10
+	} else {
+		d = 5 * time.Minute
+	}
+
+	ticker := time.NewTicker(d)
+	go func() {
+		for {
+			select {
+			case <-ticker.C:
+				logger.Debug("rekeyAllCommunities ticker fired")
+				m.rekeyAllCommunities(logger)
+				logger.Debug("rekeyAllCommunities ticker loop ended")
+			case <-m.quit:
+				ticker.Stop()
+				logger.Debug("CommunityRekeyLoop stopped")
+				return
+			}
+		}
+	}()
+}
+
+// rekeyAllCommunities attempts to rekey every community in persistence.
+// A community will be rekeyed if it meets all the following criteria:
+//   - Community.IsAdmin()
+//   - Community.Encrypted()
+//   - Community.RekeyedAt().Add(rki).Before(time.Now()) where rki is a defined rekey interval
+func (m *Messenger) rekeyAllCommunities(logger *zap.Logger) {
+	// Determine the rekey interval, if the value is not set as a property of m.communitiesManager
+	// default to one hour
+	// TODO in future perhaps have a community level rki rather than a global rki
+	var rki time.Duration
+	if m.communitiesManager.RekeyInterval == 0 {
+		rki = time.Hour
+	} else {
+		rki = m.communitiesManager.RekeyInterval
+	}
+
+	// Get and loop over all communities in persistence
+	cs, err := m.Communities()
+	if err != nil {
+		logger.Error("error getting communities", zap.Error(err))
+		return
+	}
+	for _, c := range cs {
+		if c.IsAdmin() && c.Encrypted() && c.RekeyedAt().Add(rki).Before(time.Now()) {
```

Comment:
Probably does not make much sense, but: would it be ok to move the boolean checks to SQL query itself, instead of using a generic `m.Communities()` call?

Reply:
I see what you mean, that would be a much more efficient and clean.

But unfortunately a SQL query for this information is not possible for 2 reasons:

- I'm refactoring this PR to remove the concept of `rekeyedAt` and instead rely on the keyID timestampID which is an encryptor concern. (although technically we could query across multiple tables that exist in the same DB I don't know how messy that would get.)
- The encrypted state of community is stored in the description which is marshalled into bytes in the DB so isn't searchable.

Reply:
Ahhh, of course, clear now.